azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusUpdatingListener.java

```diff
@@ -0,0 +1,317 @@
+   * flow status is in a finalized state in the Db. <br>
+   * This indirectly impacts this listener implementation as occasionally more than one thread
+   * could try to update the state of the same flow in db. Although it's not any worse than how
+   * the rest of Azkaban already behaves, we should fix the behavior at least during the
+   * finalization of flows. One way of achieving this is to add a utility method to atomically
+   * test-and-finalize a flow from a non-final to a failed state.
+   *
+   * @implNote Flow status check and update is not atomic, details above.
+   * @param event
+   * @param expectedStatusMatcher
+   * @return
+   */
+  private azkaban.executor.Status compareAndFinalizeFlowStatus(AzPodStatusMetadata event,
+      Function<Status, Boolean> expectedStatusMatcher) {
+    requireNonNull(expectedStatusMatcher, "flow matcher must not be null");
+
+    int executionId = Integer.parseInt(event.getFlowPodMetadata().get().getExecutionId());
+    ExecutableFlow executableFlow = null;
+    try {
+      executableFlow = executorLoader.fetchExecutableFlow(executionId);
+    } catch (ExecutorManagerException e) {
+      String message = format("Exception while fetching executable flow for pod %s",
+          event.getPodName());
+      logger.error(message, e);
+      throw new AzkabanWatchException(message, e);
+    }
+    checkState(executableFlow != null, "executable flow must not be null");
+    azkaban.executor.Status originalStatus = executableFlow.getStatus();
+
+    if (!expectedStatusMatcher.apply(originalStatus)) {
+      logger.info(format(
+          "Flow for pod %s does not have the expected status in database and will be finalized.",
+          event.getPodName()));
+      final String reason = "Flow Pod execution was completed.";
+      ExecutionControllerUtils.finalizeFlow(executorLoader, alerterHolder, executableFlow, reason,
+          null);
+    }
+    return originalStatus;
+  }
+
+  /**
+   * Delete the the flow pod and any other related objects (such as services).
+   *
+   * @param event pod watch event
+   */
+  private void deleteFlowContainer(AzPodStatusMetadata event) {
+    logger.info("Deleting Flow Pod: " + event.getPodName());
+    try {
+      containerizedImpl.deleteContainer(
+          Integer.parseInt(
+              event.getFlowPodMetadata().get().getExecutionId()));
+    } catch (ExecutorManagerException e) {
+      String message = format("Exception while deleting flow container.");
+      logger.error(message, e);
+      throw new AzkabanWatchException(message, e);
+    } catch (NumberFormatException ne) {
+      String message = format("Flow metadata execution id is not a valid integer %s",
+          event.getFlowPodMetadata().get().getExecutionId());
+      throw new AzkabanWatchException(message, ne);
+    }
+  }
+
+  @Override
+  public void onPodCompleted(AzPodStatusMetadata event) {
+    requireNonNull(event, "event must not be null");
```

Comment:
Log error silently?

Reply:
We need to exit early if the argument is null in this case. The call-stack components like AzPodDrivingListener and KubernetesWatch and designed to be resilient to any exceptions reported from this code.